enhancements/network/ip-interface-selection.yaml

```diff
@@ -0,0 +1,302 @@
+  between services are found once the change is being tested more broadly
+  it will be easy to revert.
+
+  We will also test the ordering change as much as possible before merging
+  it, but it's unlikely we can exercise it to the same degree that running
+  across hundreds of CI jobs per day will.
+
+- Currently all host services are expected to listen on the same IP and
+  interface. If at some point in the future we need host services listening
+  on multiple different interfaces, this may not work. However, because we
+  are centralizing all IP selection logic in nodeip-configuration, it should
+  be possible to extend that to handle multiple interfaces if necessary.
+
+### Drawbacks
+
+This design only considers host networking, and it's likely that in the future
+we will want a broader feature that provides an interface to configure pod
+traffic routing as well. However, if/when such a feature is implemented it
+should be able to use the same configuration interface for host services
+that deployers would use directly after this is implemented.
+
+Additionally, there are already ways to [implement traffic steering for pod
+networking.](https://youtu.be/EpbUWwjadYM) We may at some point want to
+integrate them more closely, but host networking is currently a much bigger
+pain point and worth addressing on its own.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Currently this only applies to UPI clusters deployed using platform None.
+  Do we need something similar for IPI?
+
+- In UPI deployments it is also possible to set the Node IP by manually writing
+  configuration files for Kubelet and CRIO. Trying to look for all possible
+  ways a user may have configured those services seems complex and error-prone.
+  Can we just require them to use this mechanism if they want custom IP
+  selection?
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+- What additional testing is necessary to support managed OpenShift service-based offerings?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+
+### Graduation Criteria
+
+NA
+
+#### Dev Preview -> Tech Preview
+
+NA
+
+#### Tech Preview -> GA
+
+NA
+
+#### Removing a deprecated feature
+
+NA
+
+### Upgrade / Downgrade Strategy
+
+NA
+
+### Version Skew Strategy
+
+From version to version the selection process must remain consistent in order
+to avoid IPs and interfaces changing. As a result, version skew should not be
+a problem.
+
+### Operational Aspects of API Extensions
+
+NA
+
+#### Failure Modes
+
+NA
```

Comment:
This might not be relevant, but a common failure mode is when NetworkManager fails to activate `bond0` due to DHCP issues or mis-config and the node is presented with multiple slaves interfaces in the same subnet each with their own DHCP lease and the node has multiple IPs for the first time.

There is also the case during `fail_over_mac=1` where the `bond0` MAC will change and potentially get a different DHCP lease.

These are both maybe DHCP network design issues.

Reply:
Yeah, I think that's kind of tangential. If the IP changes out from under us that's a problem regardless of how we're selecting the IP.